### PR TITLE
Design Doc: ElasticDL on SQLFlow

### PR DESCRIPTION
**Pull request by terrytangyuan**
This is a design doc on integration with [ElasticDL](https://github.com/wangkuiyi/elasticdl).

### CODE REVIEW

File: doc/elasticdl_on_sqlflow.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
+# _Design:_ ElasticDL on sqlflow
+
+## Overview
+
+This is a design doc on integration with [ElasticDL](https://github.com/wangkuiyi/elasticdl).
```
Comment (tonyyang-svail) [inline doc/elasticdl_on_sqlflow.md:5]: Is https://github.com/wangkuiyi/elasticdl a public repo?

File: doc/elasticdl_on_sqlflow.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+SELECT
+    c1, c2, c3, c4, c5 as class
+FROM training_data
+TRAIN ElasticDLKerasClassifier
+WITH
+  optimizer = "optimizer"
+  loss = "loss"
+  eval_metrics = "eval_metrics_fn"
+  num_classes = 10
+  runtime.num_epochs = 2
```
Comment (tonyyang-svail) [inline doc/elasticdl_on_sqlflow.md]: FYI, in the case of ALPS as backend, a user would write`train.num_epochs`.

File: doc/elasticdl_on_sqlflow.md [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    * Pass additional parameters from `WITH` clause to `custom_model()`'s instantiation, such as `optimizer` and `loss`.
+    * Skip support for feature transformation functions such as `NUMERIC` or `BUCKET` in `COLUMN` clause for now as this requires additional design details and discussions on the use of feature column APIs.
+    * Pass column names, shapes, and types for features and labels to `dataset_fn`'s feature description that will be used in `tf.io.parse_single_example()`. For this MVP, column names can be obtained from `SELECT ... LABEL ...`. Each feature columns will be of shape `[1]` and of type `tf.float32` while label column is of shape `[1]` and of type `tf.int64` for classification problems and `tf.float32` for regression problems. In the future, this will be inferred from the ODPS table. An example `dataset_fn()` looks like the following:
```
Comment (tonyyang-svail) [inline doc/elasticdl_on_sqlflow.md]: SQLFlow can help with inferring data types and tensor shapes. :)

File: doc/elasticdl_on_sqlflow.md [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+### Implementation Details
+
+#### Training Job
+
+Steps:
+
+1. Based on `SELECT ... FROM ...`, read ODPS table and write it to [RecordIO](https://github.com/wangkuiyi/recordio) files, including both features and labels.
+2. Generate model definition file (e.g. [cifar10_functional_api.py](https://github.com/wangkuiyi/elasticdl/blob/develop/model_zoo/cifar10_functional_api/cifar10_functional_api.py)) that will be used in `TRAIN` clause, which includes:
+    * In model definition function e.g. `custom_model()`, we need to configure model input and output shapes correctly in `inputs = tf.keras.layers.Input(shape=<input_shape>)` and `outputs = tf.keras.layers.Dense(<num_classes>)`(based on `COLUMN ... LABEL ...`). For this MVP, users can provide `<input_shape>` and `<num_classes>` using `WITH` clause which will then get passed to the model constructor `custom_model(input_shape, num_classes)` via `--params` argument in ElasticDL high-level API. In the future, this will be inferred from the ODPS table.
```
Comment (tonyyang-svail) [inline doc/elasticdl_on_sqlflow.md]: SQLFlow can help with inferring data types and tensor shapes. :)

File: doc/elasticdl_on_sqlflow.md [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+INTO trained_elasticdl_keras_classifier;
+```
+
+### Implementation Details
+
+#### Training Job
+
+Steps:
+
+1. Based on `SELECT ... FROM ...`, read ODPS table and write it to [RecordIO](https://github.com/wangkuiyi/recordio) files, including both features and labels.
```
Comment (tonyyang-svail) [inline doc/elasticdl_on_sqlflow.md]: Where should we store the RecordIO file? Why not reading from the ODPS table directly?

File: doc/elasticdl_on_sqlflow.md [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+                return features, tf.cast(r["label"], tf.int32)
+
+        dataset = dataset.map(_parse_data)
+
+        if mode != Mode.PREDICTION:
+            dataset = dataset.shuffle(buffer_size=1024)
+        return dataset
+    ```
+    * Pass `INTO` clause to `--outputs` argument in ElasticDL high-level API.
+3. Submit ElasticDL training job via a generated ElasticDL high-level API or CLI.
```
Comment (tonyyang-svail) [inline doc/elasticdl_on_sqlflow.md]: Maybe add the sample code for the high-level API.

File: doc/elasticdl_on_sqlflow.md (hunk 5 above)
Comment (terrytangyuan) [reply doc/elasticdl_on_sqlflow.md]: For now we have to convert them to RecordIO files. They will be stored in Kubernetes persistent volumes.

File: doc/elasticdl_on_sqlflow.md (hunk 4 above)
Comment (terrytangyuan) [reply doc/elasticdl_on_sqlflow.md]: Great! Is there an example for that?

File: doc/elasticdl_on_sqlflow.md (hunk 1 above)
Comment (terrytangyuan) [reply doc/elasticdl_on_sqlflow.md:5]: Not yet.

File: doc/elasticdl_on_sqlflow.md (hunk 2 above)
Comment (terrytangyuan) [reply doc/elasticdl_on_sqlflow.md]: Isn’t that “train” redundant with the “TRAIN” clause?

Comment (terrytangyuan) [reply doc/elasticdl_on_sqlflow.md]: The goal is to differentiate with other model-specific parameters, similar to TF Estimator’s RunConfig.

File: doc/elasticdl_on_sqlflow.md [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+WITH
+  optimizer = "optimizer"
+  loss = "loss"
+  eval_metrics = "eval_metrics_fn"
+  num_classes = 10
+  runtime.num_epochs = 2
+  runtime.master_resource_request = "cpu=400m,memory=1024Mi"
+  runtime.master_resource_limit = "cpu=400m,memory=1024Mi"
+  runtime.worker_resource_request = "cpu=400m,memory=2048Mi"
+  runtime.worker_resource_limit = "cpu=1,memory=3072Mi"
```
Comment (skydoorkai) [inline doc/elasticdl_on_sqlflow.md]: It is too complicated for SQL users to specify worker/master resources. Each predefined model can have an optimal resource config to be used directly.

File: doc/elasticdl_on_sqlflow.md (hunk 4 above)
Comment (skydoorkai) [reply doc/elasticdl_on_sqlflow.md]: For Keras functional API, input layers are required. If the Keras subclass is used, no need to define the input layer.

File: doc/elasticdl_on_sqlflow.md (hunk 7 above)
Comment (ywskycn) [reply doc/elasticdl_on_sqlflow.md]: Yes, another approach is to pre-define some resource types like "small", "standard", "large". But from API side, we still provide a way for advanced users to config their tasks, right? Not sure how ALPS handles this.

File: doc/elasticdl_on_sqlflow.md (hunk 5 above)
Comment (ywskycn) [reply doc/elasticdl_on_sqlflow.md]: @tonyyang-svail Support reading directly from ODPS will be added later, after we complete the entire SQLFlow->ElasticDL work.

File: doc/elasticdl_on_sqlflow.md [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+#### Training Job Submission
+
+```sql
+SELECT 
+    c1, c2, c3, c4, c5 as class
+FROM training_data
+TRAIN ElasticDLKerasClassifier
+WITH
+  optimizer = "optimizer"
```
Comment (weiguoz) [inline doc/elasticdl_on_sqlflow.md]: Ending with a comma.

File: doc/elasticdl_on_sqlflow.md (hunk 7 above)
Comment (terrytangyuan) [reply doc/elasticdl_on_sqlflow.md]: Just checked with ALPS folks. They handle this types of configurations in WITH clause as well.

**Comment by terrytangyuan** [comment]
Thanks everyone. I've addressed all the comments. Please take another look when you get a chance.

File: doc/elasticdl_on_sqlflow.md (hunk 4 above)
Comment (tonyyang-svail) [reply doc/elasticdl_on_sqlflow.md]: > Great! Is there an example for that?

@terrytangyuan the data type inference logic is scattered around in the code. I apologize for that.

For example, to determine the label type, we have

https://github.com/sql-machine-learning/sqlflow/blob/d873bc994350493a599e58a7cc5dc519d09e58e5/sql/codegen.go#L190-L205

To determine the input tensor shape in the case of dense tensor like "1,2,3,4". We select one row and split the cell by delimiter, then count the number of elements to determine the shape.

https://github.com/sql-machine-learning/sqlflow/blob/d873bc994350493a599e58a7cc5dc519d09e58e5/sql/codegen_alps.go#L658-L661

Currently, the type and shape inference logic is not unified across different `codegen_*.go`. SQLFlow team is working on unifying them and put the inferred result into a struct like`featureMeta`.

https://github.com/sql-machine-learning/sqlflow/blob/d873bc994350493a599e58a7cc5dc519d09e58e5/sql/codegen.go#L47-L53

This effort may take several weeks.

What is the first use case of EDL on SQLFlow?

**Review by tonyyang-svail — APPROVED** [review]
@terrytangyuan Thanks for submitting this design doc. LGTM.
